File: python/fenwick_tree.py
# TLE -> update, query 數值錯了, < 0
class BIT:
    def __init__(self, n):
        # valid input: [0, n]
        self.sums = [0] * (n+3)
    
    def update(self, i, delta):
        # update >=i with delta
        i += 2
        while i < len(self.sums):
            self.sums[i] += delta
            i += i & (-i)
    def _query(self, i):
        # get (0, i]
        i += 2
        res = 0
        while i > 0:
            res += self.sums[i]
            i -= i & (-i)
        return res
    
    def query(self, i, j =None):
        # [0, i] or [i, j]
        if j == None:
            return self._query(i)
        else:
            assert j >= i
            return self._query(j) - self._query(i-1)
        
    def update_interval(self, a, b, delta):
        # update [b, a]
        self.update(a+1, -delta)
        self.update(b, delta)
    
    
class Solution:
    def goodTriplets(self, A, B):
        d = {x: i for i, x in enumerate(A)}
        n = len(A)
        arr = [d[B[i]] for i in range(n)]

        BIT1, BIT2, ans = BIT(n), BIT(n), 0
        for i in arr:
            ans += BIT2.query(i)
            BIT1.update(i + 1, 1)
            less = BIT1.query(i)
            BIT2.update(i + 1, less)
        return ans

File: c++/fenwick_tree.cpp
template<typename T>
class BIT{
public:
    int _max;
    int offset = 2; // -> update(-1, v), query(-1, v)
    vector<T> c;
    BIT(int __max){
        _max = __max + offset;
        c.reserve(_max);
        for(int i = 0;i<_max;i++){
            c[i] = e();
        }
    }
    void update(int i, int v){
        i += offset;
        while(i < _max){
            c[i] = op(c[i], v);
            i += i & (-i);
        }
    }
    T query(int i) {
        // [0, i)
        T ans = 0;
        i = i + offset;
        i--;
        while(i > 0){
            ans = op(ans, c[i]);
            i -= i & (-i);
        }
        return ans;
    }
    T e(){return T(0);}
    T op(T a, T b){return a+b;}
};